🚀 2단계 - 수강신청(도메인 모델)

khc41 opened the pull request:
안녕하세요!
사다리게임 하고오느라 구현이 늦었습니다..!
어려워서 제대로 짰는지 모르겠습니다..🥲
PR 부탁드립니다😃

## Step1. 피드백 구현
- [X] Readme 에 체크리스트 적용
- [X] DeleteHistory에 정적 팩토리 메소드 패턴 구현
- [X] 접근 제어자 추가 (private)

## Step1. 리팩토링
- [X] 테스트 통과하게 DeleteHistory add 순서 변경
- [X] Answers에 있던 validation 체크 로직 answer 내부로 이동
- [X] DeleteHistory 생성, delete 메소드 분리 (CQRS 패턴)
- [X] question에서 title, body 분리

## Step2. 기능 구현
- [X] ImageSize validation 하는 클래스 구현
- [X] 이미지 확장자 관리 enum 구현
- [X] 강의 시작, 종료일 관리 클래스 구현
- [X] 이미지 관리 클래스 구현
- [X] 유료, 무료 강의 구현
- [X] 과정 구현
- [X] 결제 정보 구현
- [X] 결제정보 담는 일급 컬렉션 구현

heechul-kang committed:
refactor: DeleteHistory에 정적 팩토리 메소드 패턴 구현

heechul-kang committed:
refactor: 접근제어자 추가

heechul-kang committed:
refactor: 테스트 통과하게 DeleteHistory add 순서 변경 / 정적팩토리메소드 패턴 구현

heechul-kang committed:
feat: ImageSize validation 하는 클래스 구현

heechul-kang committed:
feat: 이미지 확장자 관리 enum 구현

heechul-kang committed:
feat: 강의 시작, 종료일 관리 클래스 구현

heechul-kang committed:
feat: 이미지 관리 클래스 구현

heechul-kang committed:
feat: 유료강의 구현

heechul-kang committed:
feat: 유료강의 정원초과

khc41 committed:
강의 후 리팩토링 부 정리

heechul-kang committed:
feat: step1 리팩토링

- Answers에 있던 validation 체크 로직 answer 내부로 이동
- DeleteHistory 생성, delete 메소드 분리 (CQRS 패턴)

heechul-kang committed:
feat: step1 리팩토링

- Answers에 있던 validation 체크 로직 answer 내부로 이동
- DeleteHistory 생성, delete 메소드 분리 (CQRS 패턴)

heechul-kang committed:
refactor: 클래스 이름 변경

heechul-kang committed:
refactor: positive number 삭제

heechul-kang committed:
feat: 유료, 무료 강의 구현

heechul-kang committed:
feat: 과정 구현

heechul-kang committed:
feat: 과정 구현

heechul-kang committed:
feat: 결제정보 담는 일급 컬렉션 구현

mskangg commented on README.md:
```diff
@@ -18,13 +18,48 @@
+  * 결제를 완료한 결제 정보는 payments 모듈을 통해 관리되며, 결제 정보는 Payment 객체에 담겨 반한된다.
+
+## Step2. 기능 구현
+- [X] ImageSize validation 하는 클래스 구현
+- [X] 이미지 확장자 관리 enum 구현
+- [X] 강의 시작, 종료일 관리 클래스 구현
+- [X] 이미지 관리 클래스 구현
+- [X] 유료, 무료 강의 구현
+- [X] 과정 구현
```
체크리스트 👍 
다음부턴 조금더 세부적인 기능이 나타나게 작성해보는건 어떨까요?

khc41 replied:
넵..! 다음부터는 좀 더 세부적으로 작성해보겠습니다!!
사실 설계를 고치느라 커밋할 정신이 없긴 했어요😅

mskangg commented on src/main/java/nextstep/courses/domain/Session.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    public void enroll(NsUser user) {
+        if (!RECRUITING.equals(status)) {
```
메세지를 보내어 처리할 수 있을것 같아요 :)

```suggestion
        if (!status.canEnroll()) {
```

khc41 replied:
맞네요! enum 내부에 메소드 구현했습니다!

mskangg commented on src/main/java/nextstep/courses/domain/Session.java:
```diff
@@ -0,0 +1,35 @@
+
+import static nextstep.courses.domain.SessionStatus.RECRUITING;
+
+public abstract class Session {
```
abstract 로 구현해주셨군요 👍

mskangg commented on src/main/java/nextstep/courses/domain/PaidSession.java:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    @Override
+    public void enroll(NsUser user) {
```
`유료 강의는 수강생이 결제한 금액과 수강료가 일치할 때 수강 신청이 가능하다.`

sessionFee 를 확인해야하지 않을까요?

khc41 replied:
이거를 구현 할때 정말 많은 고민이 있었어요..
기존에 메소드를 구현해서 validation을 진행했었는데, 
추상클래스로 메소드를 땡겨오는데 amount를 파라미터로 받으면, FreeSession과 PaidSession의 사용 메소드가 달라지게 되고,
부모에 메소드가 있어서 PaidSession은 enroll(user, amount)만 사용해야 하지만, PaidSession의 enroll(user)도 사용 가능하므로 원치않는 사용을 막고 싶었거든요

enroll(user)메소드를 재정의해서 
`enroll(user,0)` 이런식으로도 구현했었는데 정말 이게 맞을까 의문도 들더라구요..
그래서 추상 클래스를 분리해서 각각의 클래스로 구현도 해봤는데 결국 세션을 담는 Course에서 막히더라구요

최종적으로 생각했던 방법은 수강신청 전에 Payment쪽에서 유저의 결제정보를 가져와서 해당 강의의 결제 정보가 있으면 수강신청이 가능하게끔? 하는 방향으로 구현을 해볼려고 했었습니다.

이 문제를 해결할 좋은 방법이 있을까요..?

mskangg replied:
고민이 많으셨군요..! 👍 좋은 방향입니다 😁
지금 구조에서 파라미터를 `payment` 로 변경하는건 어떨까요?

`void enroll(Payment payment)`

결제정보 안에 유저정보도 있으니 가능할것 같아요.

mskangg commented on src/main/java/nextstep/courses/domain/SessionDate.java:
```diff
@@ -0,0 +1,20 @@
+
+import java.time.LocalDateTime;
+
+public class SessionDate {
```
👍

mskangg commented on src/main/java/nextstep/courses/domain/SessionImage.java:
```diff
@@ -0,0 +1,17 @@
+    private final SessionImageSize sessionImageSize;
+
+    public SessionImage(int volume, SessionImageType type, SessionImageSize sessionImageSize) {
+        if (volume > 1024) {
+            throw new IllegalArgumentException("이미지의 용량은 1MB를 초과할 수 없습니다.");
+        }
```
👍

mskangg commented on src/main/java/nextstep/courses/domain/SessionImageType.java:
```diff
@@ -0,0 +1,23 @@
+
+import java.util.Arrays;
+
+public enum SessionImageType {
+    GIF("gif"),
+    JPG("jpg"),
+    JPEG("jpeg"),
+    PNG("png"),
+    SVG("svg");
+    private final String extension;
```
enum 으로 잘구현해주셨네요 👍

mskangg commented on src/test/java/nextstep/courses/domain/PaidSessionTest.java:
```diff
@@ -0,0 +1,44 @@
+import static nextstep.courses.domain.SessionStatus.RECRUITING;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class PaidSessionTest {
```
`유료 강의는 수강생이 결제한 금액과 수강료가 일치할 때 수강 신청이 가능하다.`
를 검증하는 테스트코드가 없네요 🧐

khc41 replied:
추가했습니다! 
Payment로 넘기니까 좋네요 ㅎㅎ

mskangg reviewed (changes requested):
2단계 잘 구현해주셨네요! 👏 👍 
> 어려워서 제대로 짰는지 모르겠습니다..🥲

이 과정이 TDD 에서 가장 어려운 과정이긴합니다.. 🥇 
그럼에도 불구하고 정말 잘 구현해주셨어요! ㅎㅎ
몇가지 피드백 남겨드렸는데, 확인부탁드릴게요! 🙇

khc41 committed:
refactor: SessionStatus 내부에 canEnroll 메소드 구현

khc41 committed:
refactor: Payments 클래스 삭제

khc41 committed:
refactor: Session 클래스 변경

- Payments 클래스 삭제
- enroll 메소드 매개변수 Payment로 수정
- sessionId string -> long 수정
- Session 내 student에 userId 수집

khc41 committed:
refactor: Payment 내 지불한 금액 확인 메소드 추가

khc41 committed:
refactor: PaidSession 내 지불한 금액 확인 validation 추가

khc41 committed:
refactor: FreeSession 내 지불한 금액 확인 validation 추가

khc41 committed:
refactor: 안쓰는 메소드 삭제

khc41 committed:
refactor: 메소드 병합

khc41 committed:
refactor: 메소드 병합

khc41 commented:
안녕하세요!
고민했던 부분 잘 알려주셔서 감사합니다 ㅎㅎ
PR 부탁드립니다!

## Step2. 피드백 구현
- [X] SessionStatus 내부에 canEnroll 메소드 구현
- [X] Payments 클래스 삭제
- [X] enroll 메소드 매개변수 Payment로 수정 
- [X] sessionId string -> long 수정
- [X] Session 내 student에 userId 수집
- [X] Payment 내 지불한 금액 확인 메소드 추가 
- [X] PaidSession 내 지불한 금액 확인 validation 추가
- [X] FreeSession 내 지불한 금액 확인 validation 추가

mskangg commented on src/main/java/nextstep/courses/domain/PaidSession.java:
```diff
@@ -0,0 +1,31 @@
+        if (maxNumOfStudents <= students.size()) {
+            throw new IllegalStateException("수강 정원이 초과됐습니다.");
+        }
+        if (!payment.isPaid(sessionFee)) {
+            throw new IllegalStateException("지불한 수강료가 일치하지 않습니다.");
+        }
```
👍

mskangg commented on src/main/java/nextstep/courses/domain/FreeSession.java:
```diff
@@ -0,0 +1,20 @@
+
+    @Override
+    public void enroll(Payment payment) {
+        if (!payment.isPaid(0)) {
```
무료는 이렇게 처리해주셨군요 👍 

```suggestion
        if (!payment.isFree()) {
```

위처럼 메세지를 보내는것도 좋을것 같아요 :)
0원을 입력하는건 사용측이 아닌 결제정보의 책임으로 보여집니다!

khc41 replied:
0원에 대한 책임을 Payment로 옮기는 게 좋아보이네요
수정했습니다!

mskangg commented on src/test/java/nextstep/courses/domain/PaidSessionTest.java:
```diff
@@ -0,0 +1,50 @@
+    @Test
+    @DisplayName("결제한 금액과 수강료가 일치하지 않을 때, 예외를 발생시킨다.")
+    void 결제금액_불일치() {
+        assertThatThrownBy(() -> ps4.enroll(p1))
```
👍

mskangg commented on src/test/java/nextstep/courses/domain/PaidSessionTest.java:
```diff
@@ -0,0 +1,50 @@
+                .isInstanceOf(IllegalStateException.class);
+    }
+
+    @Test
+    @DisplayName("유료강의의 처음 상태가 정원초과일 경우, 예외를 발생시킨다.")
+    void 유료강의_정원초과() {
+        assertThatThrownBy(() -> new PaidSession(1L, sd1, si1, END, List.of(1L), 0, 25000))
+                .isInstanceOf(IllegalArgumentException.class);
+    }
```
이 테스트는 `유료강의의 처음 상태가 정원초과일 경우` 가 중요한 사항이죠!
그런데도 `1L, sd1, si1, END, List.of(1L), 0, 25000` 이러한 알수 없는 값들이 들어가있네요 🤔

테스트코드도 가독성이 중요하다고 생각해요.
이 테스트에서 필요한건 students.size 와 남은 정원의 숫자겠죠?

픽스쳐 관련해서 도움이 될만한 링크 남겨드릴게요!
좋은 인사이트 얻어가셨으면 좋겠습니다

https://www.arhohuttunen.com/test-data-builders/
https://jojoldu.tistory.com/611

khc41 replied:
Fixture 적용을 static 메소드로 하니까 확실히 깔끔해지네요!
감사합니다!

mskangg reviewed (approved):
피드백 반영 잘해주셨네요! 👍 
추가적으로 소소하게 피드백 남겨드렸는데, 확인부탁드릴게요!
그럼 3단계도 화이팅 🔥
